#Getting data

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [2]:
dfg = pd.read_csv('/content/cleaned_tweets_10000 (Gene).csv')

In [3]:
dfg

id      tweet_id                 date_time     author_id  \
0         1  1.220000e+18  2020-01-21T19:50:12.000Z  7.143272e+07   
1         2  1.220000e+18  2020-01-21T20:31:19.000Z  1.597135e+08   
2         3  1.220000e+18  2020-01-21T21:11:52.000Z  2.915478e+07   
3         4  1.220000e+18  2020-01-21T21:18:16.000Z  5.653205e+07   
4         5  1.220000e+18  2020-01-21T21:48:38.000Z  7.787764e+07   
...     ...           ...                       ...           ...   
9995   9996  1.500000e+18  2022-02-24T18:53:50.000Z  5.038248e+07   
9996   9997  1.500000e+18  2022-02-24T19:44:18.000Z  2.130565e+07   
9997   9998  1.500000e+18  2022-02-24T20:25:13.000Z  3.355189e+09   
9998   9999  1.500000e+18  2022-02-24T20:28:33.000Z  1.470000e+18   
9999  10000  1.500000e+18  2022-02-24T21:33:19.000Z  3.240659e+09   

              place_id               place_name        region  \
0     3797791ff9c0e4c6         Toronto, Ontario       Ontario   
1     6a6d896ba1cb5dc4        Edmonton, Alberta       Alberta   
2     4bb41f9d86e16416  Saskatoon, Saskatchewan  Saskatchewan   
3     53504716d445dcad         Calgary, Alberta       Alberta   
4     48ad377ff2ff9e51        Kingston, Ontario       Ontario   
...                ...                      ...           ...   
9995  71bdc845bc7609c7       Woodstock, Ontario       Ontario   
9996  3797791ff9c0e4c6         Toronto, Ontario       Ontario   
9997  3797791ff9c0e4c6         Toronto, Ontario       Ontario   
9998  0811cf61cd9ea52f       Winnipeg, Manitoba      Manitoba   
9999  38d5974e82ed1a6c          Ottawa, Ontario       Ontario   

                                             tweet_text  \
0     I struggle to swallow one pill and seniors are...   
1     @dowchar Haha. I often say have a good weekend...   
2     Today a student told me that @Twitter is for t...   
3     Why do elderly people tend to over react so mu...   
4     Elderly female transported to hospital with a ...   
...                                                 ...   
9995  Be part of the movement from hallway care to h...   
9996  @cllrainslie Vaccines are great for the elderl...   
9997  This is no joke. I have gone to several stores...   
9998  Winnipeg seniors, wya?!! \nI’m doin...   
9999  @JosephConwell7 @Reuters I live here &amp; the...   

                                           cleaned_text  \
0     I struggle to swallow one pill and seniors are...   
1     Haha. I often say have a good weekend to senio...   
2     Today a student told me that is for the elderl...   
3     Why do elderly people tend to over react so mu...   
4     Elderly female transported to hospital with a ...   
...                                                 ...   
9995  Be part of the movement from hallway care to h...   
9996  Vaccines are great for the elderly and / or th...   
9997  This is no joke. I have gone to several stores...   
9998  Winnipeg seniors, wya?!! Im doing Mini session...   
9999  I live here and these far right bully boys pic...   

                               hashtags  \
0                                    []   
1                                    []   
2                                    []   
3                                    []   
4     ['ygk', 'Kingston', 'RoadSafety']   
...                                 ...   
9995        ['Woodstock', 'Healthcare']   
9996                                 []   
9997                                 []   
9998        ['winnipeg', 'collegegrad']   
9999                                 []   

                                          cleaned_tweet clean_issue clean_way  \
0     I struggle to swallow one pill and seniors are...         NaN       NaN   
1     Haha. I often say have a good weekend to senio...         NaN       NaN   
2     Today a student told me that is for the elderl...     grammar       NaN   
3     Why do elderly people tend to over react so mu...         NaN       NaN   
4     Elderly female transported to hospital with a ...    

In [4]:
texts = dfg['cleaned_tweet'].sample(n=1000, random_state=48)
texts.reset_index(drop=True, inplace=True) # resetting index so they go from 0 to 999.
texts

0      Just remember you young leaders there is a lot...
1      I get the same thing on my neck and I just hat...
2      Simply because after the elderly in long term ...
3      I understand and think its a good consideratio...
4      Even though the 20 something is will go home a...
                             ...                        
995         elderly body builder classic music breakdown
996    I am 31 turning 32 in April, so I want to say ...
997    Immigration is a huge issue in Toronto, Montre...
998    This is why National Standards matter New Brun...
999    The value Canada places on seniors is incredib...
Name: cleaned_tweet, Length: 1000, dtype: object

#BERTopic setup

In [5]:
!pip install bertopic

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 143.4/143.4 kB 3.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.2/5.2 MB 16.3 MB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 88.2/88.2 kB 12.5 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 86.0/86.0 kB 11.2 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 7.4/7.4 MB 37.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.3/1.3 MB 43.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 268.8/268.8 kB 30.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.1/1.1 MB 42.4 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 7.8/7.8 MB 66.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━

In [6]:
from bertopic import BERTopic
from umap import UMAP
from sklearn.feature_extraction.text import CountVectorizer
from bertopic.vectorizers import ClassTfidfTransformer
from bertopic.representation import KeyBERTInspired
from bertopic.representation import MaximalMarginalRelevance

In [7]:
# to get the same result every time
umap_model = UMAP(
    #n_neighbors=3,
    #n_components=5,
    #min_dist=0.0,
    #metric='cosine',
    random_state=42)

In [8]:
vectorizer_model = CountVectorizer(
    min_df=0.08, # ignore terms that appear in less than 8% of the documents
    max_df = 0.15, # ignore terms that appear in more than 15% of the documents. Hopefully we'll get rid of (old, elder, senior, etc.)
    ngram_range=(1, 2), # allowing to get values like "British Columbia" instead of "British", and "Columbia".
    stop_words="english" # removing english stop words
    )

In [9]:
ctfidf_model = ClassTfidfTransformer(
    bm25_weighting=True, # At smaller datasets, this variant can be more robust to stop words that appear in your data
    reduce_frequent_words=True # To further reduce these frequent words
    )

In [10]:
# The main representation of a topic
main_representation = KeyBERTInspired()

# Additional ways of representing a topic
add_representation = [MaximalMarginalRelevance(diversity=.1)]

# Add all models together to be run in a single `fit`
representation_model = {
   "Main": main_representation,
   "Aspect1":  add_representation
}

In [11]:
seed_topic_list = [
    ['candidate', 'election', 'government', 'law', 'minister', 'president', 'senator', 'state', 'vote'],
    ['coronavirus', 'covid19', 'dose', 'hospital', 'immunity', 'nurse', 'outbreak', 'treatment', 'vaccine', 'wellness'],
    ['celebrate', 'ceremony', 'christmas', 'economy', 'hate', 'immigration', 'indian', 'labour', 'poverty', 'prejudice', 'protests', 'retired', 'transgender', 'veteran', 'wage', 'wheelchair'],
    ['athlete', 'club', 'hockey', 'jersey', 'player', 'team']
    ]

# BERTopic (1000 texts)

In [35]:
model = BERTopic(
    umap_model = umap_model,
    vectorizer_model=vectorizer_model,
    ctfidf_model=ctfidf_model,
    representation_model=representation_model,
    seed_topic_list=seed_topic_list, # to guide the topics towards the seed_topic_list
    nr_topics = 10
    )

In [36]:
topics, probs = model.fit_transform(texts)

In [14]:
model.get_topic_info()

Topic  Count                                               Name  \
0     -1    358  -1_transgender elders_transgender group_christ...   
1      0    169  0_housing seniors_make seniors_doing seniors_s...   
2      1    124  1_seniors flu_vaccinated people_vaccinated vac...   
3      2    124  2_seniors lodge_community centre_local seniors...   
4      3     79  3_politics canada_canada emergency_canada pens...   
5      4     59  4_weary years_remember remembrance_grow old_ag...   
6      5     31  5_indigenous elders_reconciliation indigenous_...   
7      6     31  6_seniors play_playing senior_hockey team_mens...   
8      7     13  7_protestors_know protestors_protestors idea_e...   
9      8     12  8_martin elders_old tv_old remember_younger elder   

                                      Representation  \
0  [transgender elders, transgender group, christ...   
1  [housing seniors, make seniors, doing seniors,...   
2  [seniors flu, vaccinated people, vaccinated va...   
3  [seniors lodge, community centre, local senior...   
4  [politics canada, canada emergency, canada pen...   
5  [weary years, remember remembrance, grow old, ...   
6  [indigenous elders, reconciliation indigenous,...   
7  [seniors play, playing senior, hockey team, me...   
8  [protestors, know protestors, protestors idea,...   
9  [martin elders, old tv, old remember, younger ...   

                                             Aspect1  \
0  [transgender elders, transgender group, christ...   
1  [housing seniors, make seniors, doing seniors,...   
2  [seniors flu, vaccinated people, vaccinated va...   
3  [seniors lodge, community centre, local senior...   
4  [politics canada, canada emergency, canada pen...   
5  [weary years, remember remembrance, grow old, ...   
6  [indigenous elders, reconciliation indigenous,...   
7  [seniors play, playing senior, hockey team, me...   
8  [protestors, know protestors, protestors idea,...   
9  [martin elders, old tv, old remember, younger ...   

                                 Representative_Docs  
0  [Bronx lesbian, gay, bisexual, and transgender...  
1  [As long as she does not escape the enclosure,...  
2  [COVID 19 test is only for elderly and comorbi...  
3  [Thanks to my friend Tony Lantorno and his tea...  
4  [He has survived off of the $10m surplus left ...  
5  [They shall grow not old, as we that are left ...  
6  [The 11 Regional Municipality of Wood Buffalo ...  
7  [Go check out my highlights playing up with th...  
8  [This is a attempted MURDER or assault with we...  
9  [Tonight asks, 'Is old age the secret to happi...

In [15]:
model.get_document_info(texts)

Document  Topic  \
0    Just remember you young leaders there is a lot...      0   
1    I get the same thing on my neck and I just hat...     -1   
2    Simply because after the elderly in long term ...      1   
3    I understand and think its a good consideratio...     -1   
4    Even though the 20 something is will go home a...      1   
..                                                 ...    ...   
995       elderly body builder classic music breakdown     -1   
996  I am 31 turning 32 in April, so I want to say ...      0   
997  Immigration is a huge issue in Toronto, Montre...     -1   
998  This is why National Standards matter New Brun...      3   
999  The value Canada places on seniors is incredib...      3   

                                                  Name  \
0    0_housing seniors_make seniors_doing seniors_s...   
1    -1_transgender elders_transgender group_christ...   
2    1_seniors flu_vaccinated people_vaccinated vac...   
3    -1_transgender elders_transgender group_christ...   
4    1_seniors flu_vaccinated people_vaccinated vac...   
..                                                 ...   
995  -1_transgender elders_transgender group_christ...   
996  0_housing seniors_make seniors_doing seniors_s...   
997  -1_transgender elders_transgender group_christ...   
998  3_politics canada_canada emergency_canada pens...   
999  3_politics canada_canada emergency_canada pens...   

                                        Representation  \
0    [housing seniors, make seniors, doing seniors,...   
1    [transgender elders, transgender group, christ...   
2    [seniors flu, vaccinated people, vaccinated va...   
3    [transgender elders, transgender group, christ...   
4    [seniors flu, vaccinated people, vaccinated va...   
..                                                 ...   
995  [transgender elders, transgender group, christ...   
996  [housing seniors, make seniors, doing seniors,...   
997  [transgender elders, transgender group, christ...   
998  [politics canada, canada emergency, canada pen...   
999  [politics canada, canada emergency, canada pen...   

                                               Aspect1  \
0    [housing seniors, make seniors, doing seniors,...   
1    [transgender elders, transgender group, christ...   
2    [seniors flu, vaccinated people, vaccinated va...   
3    [transgender elders, transgender group, christ...   
4    [seniors flu, vaccinated people, vaccinated va...   
..                                                 ...   
995  [transgender elders, transgender group, christ...   
996  [housing seniors, make seniors, doing seniors,...   
997  [transgender elders, transgender group, christ...   
998  [politics canada, canada emergency, canada pen...   
999  [politics canada, canada emergency, canada pen...   

                                   Representative_Docs  \
0    [As long as she does not escape the enclosure,...   
1    [Bronx lesbian, gay, bisexual, and transgender...   
2    [COVID 19 test is only for elderly and comorbi...   
3    [Bronx lesbian, gay, bisexual, and transgender...   
4    [COVID 19 test is only for elderly and comorbi...   
..                                                 ...   
995  [Bronx lesbian, gay, bisexual, and transgender...   
996  [As long as she does not escape the enclosure,...   
997  [Bronx lesbian, gay, bisexual, and transgender...   
998  [He has survived off of the $10m surplus left ...   
999  [He has survived off of the $10m surplus left ...   

                                           Top_n_words  Probability  \
0    housing seniors - make seniors - doing seniors...     0.861627   
1    transgender elders - transgender group - chris...     0.000000   
2    seniors flu - vaccinated people - vaccinated v...     0.533131   
3    transgender elders - transgender group - chris...     0.000000   
4    seniors flu - vaccinated people - vaccinated v...     0.651695   
..                                                 

## Hierarchical Topics

In [16]:
hierarchical_topics = model.hierarchical_topics(texts)

100%|██████████| 8/8 [00:02<00:00,  2.88it/s]


In [17]:
model.visualize_hierarchy(hierarchical_topics=hierarchical_topics)

In [18]:
tree = model.get_topic_tree(hierarchical_topics)

In [19]:
print(tree)

.
├─weary years_remember remembrance_grow old_age shall_shall weary
│    ├─people elder_martin elders_read elders_protestors idea_protestors
│    │    ├─■──protestors_know protestors_protestors idea_end stigma_morons racists ── Topic: 7
│    │    └─■──martin elders_old tv_old remember_younger elder_read elders ── Topic: 8
│    └─weary years_remember remembrance_grow old_age shall_shall weary
│         ├─weary years_remember remembrance_grow old_age shall_shall weary
│         │    ├─■──weary years_remember remembrance_grow old_age shall_shall weary ── Topic: 4
│         │    └─■──indigenous elders_reconciliation indigenous_politics indigenous_survivors elders_really indigenous ── Topic: 5
│         └─■──seniors play_playing senior_hockey team_mens team_play seniority ── Topic: 6
└─remember remembrance_weary years_years condemn_grow old_shall weary
     ├─remember remembrance_weary years_years condemn_grow old_shall weary
     │    ├─■──seniors lodge_community centre_local seniors_ontar

# BERT (all texts)

In [20]:
dfg['cleaned_tweet']

0       I struggle to swallow one pill and seniors are...
1       Haha. I often say have a good weekend to senio...
2       Today a student told me that is for the elderl...
3       Why do elderly people tend to over react so mu...
4       Elderly female transported to hospital with a ...
                              ...                        
9995    Be part of the movement from hallway care to h...
9996    Vaccines are great for the elderly and or the ...
9997    This is no joke. I have gone to several stores...
9998    Winnipeg seniors, where you at?! I Am doing Mi...
9999    I live here and these far right bully boys pic...
Name: cleaned_tweet, Length: 10000, dtype: object

In [21]:
model_all_texts = BERTopic(
    umap_model = umap_model,
    vectorizer_model=vectorizer_model,
    ctfidf_model=ctfidf_model,
    representation_model=representation_model,
    seed_topic_list=seed_topic_list, # to guide the topics towards the seed_topic_list
    )

In [22]:
topics_all, probs_all = model_all_texts.fit_transform(dfg['cleaned_tweet'])

In [23]:
model_all_texts.get_topic_info()

Topic  Count                                               Name  \
0       -1   4206  -1_seniors care_support seniors_seniors homes_...   
1        0    730  0_seniors vulnerable_elderly vulnerable_covid1...   
2        1    393            1_ceremony_celebrating_honour_celebrate   
3        2    247                          2_evil_hate_masks_chinese   
4        3    228                        3_bus_mobility_road_walking   
..     ...    ...                                                ...   
115    114     11  114_elder care_seniors homes_care facilities_l...   
116    115     10                  115_seniority_chair_details_grand   
117    116     10  116_edmonton international_calgary_donations_i...   
118    117     10  117_prime minister_responsibility_leadership_v...   
119    118     10           118_meds_medicine_people seniors_doctors   

                                        Representation  \
0    [seniors care, support seniors, seniors homes,...   
1    [seniors vulnerable, elderly vulnerable, covid...   
2    [ceremony, celebrating, honour, celebrate, chi...   
3    [evil, hate, masks, chinese, protecting, justi...   
4    [bus, mobility, road, walking, people seniors,...   
..                                                 ...   
115  [elder care, seniors homes, care facilities, l...   
116  [seniority, chair, details, grand, celebrate, ...   
117  [edmonton international, calgary, donations, i...   
118  [prime minister, responsibility, leadership, v...   
119  [meds, medicine, people seniors, doctors, near...   

                                               Aspect1  \
0    [seniors care, support seniors, seniors homes,...   
1    [seniors vulnerable, elderly vulnerable, covid...   
2    [ceremony, celebrating, honour, celebrate, chi...   
3    [evil, hate, masks, chinese, protecting, justi...   
4    [bus, mobility, road, walking, people seniors,...   
..                                                 ...   
115  [elder care, seniors homes, care facilities, l...   
116  [seniority, chair, details, grand, celebrate, ...   
117  [edmonton international, calgary, donations, i...   
118  [prime minister, responsibility, leadership, v...   
119  [meds, medicine, people seniors, doctors, near...   

                                   Representative_Docs  
0    [Totally heartbreaking. One of the reasons why...  
1    [Seniors in Winnipeg. are finally realizing th...  
2    [Thank you to Elder Cindy Crow for the ceremon...  
3    [What about the black hooded Anti-fascism who ...  
4    [Would it be possible to have the 56 Community...  
..                                                 ...  
115  [Edging just ahead is the Kitikmeot community ...  
116  [Tonight we celebrate our graduating seniors a...  
117  [Our socktober goal was 10,000 pairs of socks....  
118  [I think most see Our premier as the Captain o...  
119  [Absolutely. Newfoundland and Labrador have th...  

[120 rows x 6 columns]

## Hierarchical Topics

In [24]:
hierarchical_topics = model_all_texts.hierarchical_topics(dfg['cleaned_tweet'])

100%|██████████| 118/118 [00:39<00:00,  2.98it/s]


In [25]:
model_all_texts.visualize_hierarchy(hierarchical_topics=hierarchical_topics)

In [26]:
tree = model_all_texts.get_topic_tree(hierarchical_topics)

In [27]:
print(tree)

.
├─elderly vulnerable_elderly disabled_vulnerable elderly_benefit_children elderly
│    ├─alberta legislative_democratic party_canadian politics_albertans_new democratic
│    │    ├─alberta legislative_albertans_democratic party_democratic_new democratic
│    │    │    ├─voters_voting_election_voted_democratic
│    │    │    │    ├─■──prime minister_responsibility_leadership_voted_privilege ── Topic: 117
│    │    │    │    └─■──voters_voting_election_voted_democratic ── Topic: 109
│    │    │    └─alberta legislative_albertans_democratic party_new democratic_calgary
│    │    │         ├─alberta legislative_albertans_democratic party_new democratic_calgary
│    │    │         │    ├─■──alberta legislative_albertans_canadian politics_calgary_canadas ── Topic: 84
│    │    │         │    └─■──alberta legislative_albertans_democratic party_new democratic_calgary ── Topic: 54
│    │    │         └─alberta legislative_canadian politics_albertans_quebec_election
│    │    │              ├─

# Metrics

##Coherence score
https://github.com/MaartenGr/BERTopic/issues/469

In [28]:
from gensim import corpora
from gensim.corpora import Dictionary
from gensim.models.coherencemodel import CoherenceModel

In [29]:
# Extract vectorizer and analyzer from BERTopic
vectorizer = model.vectorizer_model
analyzer = vectorizer.build_analyzer()

In [39]:
tokens = [analyzer(text) for text in texts]
dictionary = corpora.Dictionary(tokens)
corpus = [dictionary.doc2bow(token) for token in tokens]
topic_words = [[words for words, _ in model.get_topic(topic)] for topic in range(len(set(topics))-1)]

In [40]:
coherence_model = CoherenceModel(
    texts=tokens,
    dictionary=dictionary,
    corpus=corpus,
    topics = topic_words,
    coherence='c_v')

In [41]:
coherence_score = coherence_model.get_coherence()
print("Coherence Score:", coherence_score)

Coherence Score: 0.49453491954325685


The coherence score is a measure of how coherent the topics generated by the BERTopic model are. A coherence score typically ranges from 0 to 1, where higher values indicate more coherent and meaningful topics.

The coherence score of 0.4945 suggests that the topics generated by the BERTopic model are moderately coherent. A score around 0.5 is generally considered reasonable, but it can vary depending on the field and dataset.

Since this is unsupervised learning, we always need human judgement to evaluate the output of our models.

In [42]:
coherence_per_topic = coherence_model.get_coherence_per_topic()
coherence_per_topic

[0.6418693920823039,
 0.6018813567397672,
 0.4968129456416811,
 0.47032449762631395,
 0.7859540055417047,
 0.3691588078622665,
 0.39164210472715355,
 0.37673645943460204,
 0.3164347062335187]

As explained before, the higher the score, the more coherent it is. Topics 1, 2, and 5 have relatively high coherence scores (0.6419, 0.6019, and 0.7860, respectively), indicating that they are well-formed and coherent topics. We could focus on these well-formed topics; consider them well defined topics within our collection of tweets. In the other hand, those topics with low values might need revisions and/or further tuning of the model.

Topic modeling is an exploratory process, and achieving the most coherent and relevant topics may require fine-tuning and experimentation. It's essential to achieve a balance between coherence and topic diversity while ensuring the topics are relevant. Furthermore, we must always manually inspect the tweets associated to each topic to ensure the topic is relevant.


##Coherence score for different models

In [43]:
# create a dataframe to store the results
coherence_results = pd.DataFrame(columns=["Run", "NumTopics", "Count-1", "Coherence", "CoherencePerTopic"])

In [44]:
coherence_results

Empty DataFrame
Columns: [Run, NumTopics, Count-1, Coherence, CoherencePerTopic]
Index: []

In [45]:
# Run a few times and store the results
for i in range(8,16):
  try:
    # BERTopic model
    model = BERTopic(umap_model = umap_model,
                    vectorizer_model=vectorizer_model,
                    ctfidf_model=ctfidf_model,
                    representation_model=representation_model,
                    seed_topic_list=seed_topic_list, # to guide the topics towards the seed_topic_list
                    nr_topics = i # to variate the number of topics in each run
      )

    # Fit BERTopic and get the topics
    topics, probs = model.fit_transform(texts)

    # Get the number of topics and count of tweets in topic -1
    num_topics = len(model.get_topic_freq())
    count_minus_1 = topics.count(-1)

    # Get coherence per model and per model.topic
    ### Extract vectorizer and analyzer from BERTopic
    vectorizer = model.vectorizer_model
    analyzer = vectorizer.build_analyzer()

    ### Extract features for Topic Coherence evaluation
    tokens = [analyzer(text) for text in texts]
    dictionary = corpora.Dictionary(tokens)
    corpus = [dictionary.doc2bow(token) for token in tokens]
    topic_words = [[words for words, _ in model.get_topic(topic)] for topic in range(len(set(topics))-1)]

    ### Evaluate Coherence
    coherence_model = CoherenceModel(texts=tokens,
                                    dictionary=dictionary,
                                    corpus=corpus,
                                    topics = topic_words,
                                    coherence='c_v')
    coherence_score = coherence_model.get_coherence()
    coherence_per_topic = coherence_model.get_coherence_per_topic()

    # Append the results to the DataFrame
    coherence_results = coherence_results.append({
        "Run": i ,
        "NumTopics": num_topics,
        "Count-1": count_minus_1,
        "Coherence": coherence_score,
        "CoherencePerTopic": coherence_per_topic
        }, ignore_index=True)

  except ValueError as e:
        print(f"Skipping iteration due to error: {e}")
        continue

Skipping iteration due to error: After pruning, no terms remain. Try a lower min_df or a higher max_df.


In [46]:
coherence_results

Run NumTopics Count-1  Coherence  \
0   8         8     358   0.503142   
1   9         9     358   0.468917   
2  10        10     358   0.494535   
3  11        11     358   0.456131   
4  12        12     358   0.442103   
5  14        14     358   0.481102   
6  15        14     358   0.481102   

                                   CoherencePerTopic  
0  [0.41271908738620056, 0.597418883582467, 0.641...  
1  [0.41271908738620056, 0.6018813567397672, 0.49...  
2  [0.6418693920823039, 0.6018813567397672, 0.496...  
3  [0.6418693920823039, 0.6018813567397672, 0.496...  
4  [0.5533669453174155, 0.6018813567397672, 0.496...  
5  [0.5530611111088364, 0.43591143065383847, 0.57...  
6  [0.5530611111088364, 0.43591143065383847, 0.57...

In [50]:
for coherence_score in coherence_results['CoherencePerTopic']:
    formatted_scores = ["{:.2f}".format(score) for score in coherence_score]
    print(formatted_scores)

['0.41', '0.60', '0.64', '0.79', '0.39', '0.38', '0.32']
['0.41', '0.60', '0.50', '0.79', '0.37', '0.39', '0.38', '0.32']
['0.64', '0.60', '0.50', '0.47', '0.79', '0.37', '0.39', '0.38', '0.32']
['0.64', '0.60', '0.50', '0.47', '0.60', '0.37', '0.39', '0.38', '0.32', '0.30']
['0.55', '0.60', '0.50', '0.47', '0.60', '0.37', '0.39', '0.39', '0.38', '0.32', '0.30']
['0.55', '0.44', '0.58', '0.40', '0.62', '0.36', '0.57', '0.52', '0.49', '0.56', '0.40', '0.40', '0.36']
['0.55', '0.44', '0.58', '0.40', '0.62', '0.36', '0.57', '0.52', '0.49', '0.56', '0.40', '0.40', '0.36']


The above data frame shows that the models with fewer topics tends to have more coherence.

As mentioned before, topic modeling is an unsupervised learning technique so the results can be ambiguous, difficult to interpret, and they are sensitive to the data fed.